Regression in interaction between lifetime parameter and defaulted type parameter

The [`mpi` crate](https://github.com/bsteinb/rsmpi) contains (approximately) the following code that used to work before nightly-2018-03-02:

```rust
use std::cell::Cell;
use std::marker::PhantomData;

pub struct Request<'a, S: Scope<'a> = StaticScope> {
    scope: S,
    phantom: PhantomData<Cell<&'a ()>>,
}

pub unsafe trait Scope<'a> {
    /// Registers a request with the scope.
    fn register(&self);

    /// Unregisters a request from the scope.
    unsafe fn unregister(&self);
}

pub struct StaticScope;

unsafe impl Scope<'static> for StaticScope {
    fn register(&self) {}

    unsafe fn unregister(&self) {}
}
```

The newest nightly compiler that successfully compiles this is:

```
rustc 1.26.0-nightly (0ff9872b2 2018-02-28)
binary: rustc
commit-hash: 0ff9872b2280009f094af0df3dcdc542cc46a5fd
commit-date: 2018-02-28
host: x86_64-apple-darwin
release: 1.26.0-nightly
LLVM version: 6.0
```

It no longer compiles on:

```
rustc 1.26.0-nightly (3eeb5a665 2018-03-01)
binary: rustc
commit-hash: 3eeb5a665e313c5b281820099e04d4c6c8188b46
commit-date: 2018-03-01
host: x86_64-apple-darwin
release: 1.26.0-nightly
LLVM version: 6.0
```

Was this not supposed to ever work or is this a regression?

Oops, it seems I completely forgot to paste the resulting error message:

```
error[E0308]: mismatched types
 --> src/lib.rs:4:1
  |
4 | / pub struct Request<'a, S: Scope<'a> = StaticScope> {
5 | |     scope: S,
6 | |     phantom: PhantomData<Cell<&'a ()>>,
7 | | }
  | |_^ lifetime mismatch
  |
  = note: expected type `Scope<'a>`
             found type `Scope<'static>`
note: the lifetime 'a as defined on the struct at 4:1...
 --> src/lib.rs:4:1
  |
4 | pub struct Request<'a, S: Scope<'a> = StaticScope> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...does not necessarily outlive the static lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `rsmpi-request-lifetime-issue`.

To learn more, run the command again with --verbose.
```

Sorry about that. As mentioned in [the original issue](https://github.com/bsteinb/rsmpi/issues/45) the problem can be worked around by not having the type parameter `S` default to `StaticScope` and instead always specify both parameters explicitly, e.g. `Request<'static, StaticScope>`

--- a/src/librustc_typeck/check/wfcheck.rs
+++ b/src/librustc_typeck/check/wfcheck.rs
@@ -423,12 +423,18 @@ fn check_where_clauses<'a, 'gcx, 'fcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
                     _ => t.super_visit_with(self)
                 }
             }
+
+            fn visit_region(&mut self, _: ty::Region<'tcx>) -> bool {
+                true
+            }
         }
         let mut param_count = CountParams { params: FxHashSet() };
-        pred.visit_with(&mut param_count);
+        let has_region = pred.visit_with(&mut param_count);
         let substituted_pred = pred.subst(fcx.tcx, substs);
-        // Don't check non-defaulted params, dependent defaults or preds with multiple params.
-        if substituted_pred.references_error() || param_count.params.len() > 1 {
+        // Don't check non-defaulted params, dependent defaults (including lifetimes)
+        // or preds with multiple params.
+        if substituted_pred.references_error() || param_count.params.len() > 1
+            || has_region {
             continue;
         }
         // Avoid duplication of predicates that contain no parameters, for example.

--- a/src/test/run-pass/defaults-well-formedness.rs
+++ b/src/test/run-pass/defaults-well-formedness.rs
@@ -27,4 +27,8 @@ trait SelfBound<T: Copy=Self> {}
 // Not even for well-formedness.
 struct WellFormedProjection<A, T=<A as Iterator>::Item>(A, T);
 
+// Issue #49344, predicates with lifetimes should not be checked.
+trait Scope<'a> {}
+struct Request<'a, S: Scope<'a> = i32>(S, &'a ());
+
 fn main() {}

